src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py

```diff
@@ -0,0 +1,287 @@
+        objs = []
+        for obj in annotations:
+            x1 = np.max((0, obj["bbox"][0]))
+            y1 = np.max((0, obj["bbox"][1]))
+            x2 = np.min((width, x1 + np.max((0, obj["bbox"][2]))))
+            y2 = np.min((height, y1 + np.max((0, obj["bbox"][3]))))
+            if obj["area"] > 0 and x2 >= x1 and y2 >= y1:
+                obj["clean_bbox"] = [x1, y1, x2, y2]
+                objs.append(obj)
+
+        num_objs = len(objs)
+
+        res = np.zeros((num_objs, 5))
+        num_seg_values = 98 if self.tight_box_rotation else 0
+        res_seg = np.ones((num_objs, num_seg_values))
+        res_seg.fill(np.nan)
+
+        for ix, obj in enumerate(objs):
+            cls = self.class_ids.index(obj["category_id"])
+            res[ix, 0:4] = obj["clean_bbox"]
+            res[ix, 4] = cls
+            if self.tight_box_rotation:
+                seg_points = [j for i in obj.get("segmentation", []) for j in i]
+                if seg_points:
+                    seg_points_c = np.array(seg_points).reshape((-1, 2)).astype(np.int)
+                    seg_points_convex = cv2.convexHull(seg_points_c).ravel()
+                else:
+                    seg_points_convex = []
+                res_seg[ix, :len(seg_points_convex)] = seg_points_convex
+
+        r = min(self.input_dim[0] / height, self.input_dim[1] / width)
+        res[:, :4] *= r
+        res_seg *= r
+
+        img_info = (height, width)
+        resized_info = (int(height * r), int(width * r))
+
+        file_name = (
+            im_ann["file_name"]
+            if "file_name" in im_ann
+            else "{:012}".format(id_) + ".jpg"
+        )
+
+        return res, res_seg, img_info, resized_info, os.path.join(self.data_dir, self.name, file_name)
+
+    def _cache_images(self):
+        logger.warning(
+            "\n********************************************************************************\n"
+            "You are using cached images in RAM to accelerate training.\n"
+            "This requires large system RAM.\n"
+            "Make sure you have 200G+ RAM and 136G available disk space for training COCO.\n"
+            "********************************************************************************\n"
+        )
+        max_h = self.input_dim[0]
+        max_w = self.input_dim[1]
+        cache_file = os.path.join(self.data_dir, f"img_resized_cache_{self.name}.array")
+        if not os.path.exists(cache_file):
+            logger.info(
+                "Caching images for the first time. This might take about 20 minutes for COCO"
+            )
+            self.imgs = np.memmap(
+                cache_file,
+                shape=(len(self.ids), max_h, max_w, 3),
+                dtype=np.uint8,
+                mode="w+",
+            )
+            from tqdm import tqdm
+            from multiprocessing.pool import ThreadPool
+
+            NUM_THREADs = min(8, os.cpu_count())
+            loaded_images = ThreadPool(NUM_THREADs).imap(
```

Comment:
i think using a Threadpool inside a distributed dataloader may cause issues.

Reply:
Not a problem. We have the dataset creation wrapped with wait_for_the_master() which waits for the caching of rank 0 to be completed. After that, the cache file already exists so this happens in rank 0 only.